add sdpa to OPT

avishaiElmakies opened the pull request:
adds SDPA to OPT model

impl inspired by ```gemma2``` and ```llama```.

part of #28005.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?
 I think @amyeroberts @fxmarty 

some notes:

- I did some refactoring to the model code. created a function ```_update_key_and_values```. the code was used by all 3 attention implementations. ```self._shape``` to ```_shape```. and moved masks logic to ```self._update_casual_mask```.
- I created a test that makes sure the generate of eager and SDPA are equivalent. the test is similar to the one ```llama```
- I seem to fail 3 implementations of common tests ```test_eager_matches_sdpa_inference_0_float16```, ```test_eager_matches_sdpa_inference_1_bfloat16```, ```test_eager_matches_sdpa_inference_2_float32```. I took inspiration from gemma, which seems to ignore those tests as well. should i ignore them as well? I think it is related also related to #32086 since my code is similar to that. Also seems the affect equivalence with flax/tf beacuse the default will become sdpa. 

would love some feedback!

Commit:
add sdpa to OPT

Commit:
chore: remove redundant whitespace in OPTDecoder class

Commit:
fixup

Commit:
bug fix

Commit:
add sdpa and attention generate test

Commit:
fixup

amyeroberts commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -60,6 +63,10 @@
 _SEQ_CLASS_EXPECTED_OUTPUT = "'LABEL_0'"
 
 
+def _shape(tensor: torch.Tensor, seq_len: int, bsz: int, num_heads: int, head_dim: int) -> torch.Tensor:
+    return tensor.view(bsz, seq_len, num_heads, head_dim).transpose(1, 2).contiguous()
```
I agree in general this is better as a function, but this breaks the pattern with other models so it's better to just leave as-is

amyeroberts commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -116,47 +123,25 @@ def __init__(
-        # get query proj
-        query_states = self.q_proj(hidden_states) * self.scaling
-        # get key, value proj
+    def _update_key_and_values(self, hidden_states, key_value_states, past_key_value, is_cross_attention, bsz):
```
I understand the motivation here (and think it's a good refactor) but like moving out `_shape` it means it diverges from other model patterns, which will make future changes more difficult, so we should leave as-is

amyeroberts reviewed (commented):
Thanks for working on adding this!

Commit:
Refactor OPTAttention forward method for improved readability and maintainability

Commit:
undo refactor for _shape and key,val states

Commit:
add OPT to doc, fixup didn't find it for some reason

Commit:
change order

amyeroberts commented:
@avishaiElmakies Similar to [my comment](https://github.com/huggingface/transformers/pull/33403#issuecomment-2340236947) for DinoV2 -- other PRs can be a good reference for how to fix. There you will see that many of the tests enforce the test model to use eager attention. This will resolve the TF/flax equivalence issues. 

For the sdpa equivalence tests, I'm not sure. As the flash attention implementation follows llama, following llama's tests for sdpa seems reasonable cc @ArthurZucker here who might now more about this

Commit:
change default attn_implemntation in testing to eager

avishaiElmakies committed:
Merge branch 'main' into spda_opt

Commit:
[run-slow] opt

Commit:
change test_eager_matches_sdpa_generate to the one llama

Commit:
Update default attention implementation in testing common

Commit:
[run-slow] opt

avishaiElmakies commented:
Hi @amyeroberts. looked at the faiiling test

changed the test ```test_eager_matches_sdpa_generate``` to match the one in ```llama```, fixed ```test_pt_tf_model_equivalence```.

the tests in the file ```test_modelling_tf_opt.py```. ```test_xla_generate_contrastive``` and ```test_xla_generate_slow```. don't seem related to this PR. they also fail in main.

the only tests which I would love some guidance on are the ```test_eager_matches_sdpa_inference```. generating seems to work the same with both implementations. from what I understand and after looking at it the behavior for both implementations is different in the inference test. as stated by #32086. fixing this might require a refactor, to make sure behaviors match.

would love some guidance.

Commit:
remove uneeded print

Commit:
[run-slow] opt

Commit:
refactor model testers to have attn_implementation="eager"

Commit:
[run-slow] opt

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_33298). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

vasqu commented:
The code looks eerily similar to bart's implementation, it might be more beneficial to let it "inherit" from bart via `# Copied from ...` statements and adjust it to fit in with barts uses of attention masks etc instead of manually figuring stuff out 👀 

Hence, it's also more appropriate to look into encoder-decoder models like bart instead of llama who are decoder only.

vasqu commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -359,9 +362,106 @@ def forward(
+            # this model uses the scaling factor in the query projection for some reason, but not in Q@K^T
+            # so we need to scale to remove scaling in SDPA to have similar results with eager.
+            # Maybe needs a change in the model to remove scaling in query projection
+            scale=1.0,
```
You can leave the scaling to SDPA and leave out the scale mul on the query states. Since it's a scalar it doesn't matter when we apply the scaling:
`(QK^T) * scale == (Q * scale)K^T` (the should be barely any differences)

That's why I'd prefer following bart as it's doing the same stuff as OPT except I'm overlooking smthn.

avishaiElmakies replied:
While you are correct about the math, it seems this way is the only way to make eager and sdpa equivalent. When I let sdpa handle scalling they were not equivalent

vasqu replied:
Interesting, but it might be also HW dependent in how much influence it does have. Not opposed to keeping it as you have it for now.

vasqu commented:
Ok #17437 seems to the cause why the copied from doesn't exist in the first place. Is there a way to just ignore a few lines since it seems a bit too much to me to remove the whole copied from then 👀

avishaiElmakies commented:
@vasqu I can take inspiration from Bart instead of llama,if that helps.

vasqu commented:
@avishaiElmakies I think we can closely follow bart, maybe even leverage a `Copied from` for the SDPA attention module. It also makes sense to follow bart regarding the attention mask creation then. It's more tested and makes our life easier :)

vasqu commented:
For the generation, they can be flaky so I'd opt (unintended word play) for up to a 1b ish model for the test. 125M is still relatively small for todays standards.

avishaiElmakies commented:
@vasqu OK, I will look into changing it more to a Bart inspiration.

I will also try to use use a larger model, i think opt has 125M, 350M and 1.3B
Do you want the 1.3B? I also hope I have the resources to test with the larger models 😅

vasqu commented:
Hopefully, the 350M suffices no need to go further than 1.3b tho. For the larger models, we can leverage the slow runs if that's necessary ;)

avishaiElmakies commented:
@vasqu OK, thanks for the guidance! I will work on this and update when I'm Done

Commit:
convert test_eager_matches_sdpa_generate to opt-350M

Commit:
bug fix when creating mask for opt

Commit:
Merge branch 'main' into spda_opt

Commit:
[run-slow] opt

avishaiElmakies commented:
@vasqu I changed the test to 350M.

I also think i found the bug thanks to bart, the 5 tests are now passing (at least on my machine)

vasqu commented:
@avishaiElmakies great to hear! Let's wait until Amy has time to review but I'll take a closer look in a bit :)

vasqu commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -359,9 +362,106 @@ def forward(
         return attn_output, attn_weights_reshaped, past_key_value
 
 
+class OPTSdpaAttention(OPTAttention):
```
I'd be more in favor of using `# Copied from ...` here tbh (as long as it doesn't break anything).

avishaiElmakies replied:
I didn't use copy because the other attn implementations are not really copy. the sdpa is also not a perfect copy as well

vasqu replied:
Ah ok, makes sense. If there was an option to disable copying for a few lines, it'd favor that more. For now, I think it's ok to leave it like this.

vasqu commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -359,9 +362,106 @@ def forward(
+        output_attentions: bool = False,
+        position_ids: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        if output_attentions:
```
It also needs a check on `layer_head_mask`, i.e., check if it is not None.

See ref:
https://github.com/huggingface/transformers/blob/78b2929c0554b79e0489b451ce4ece14d265ead2/src/transformers/models/bart/modeling_bart.py#L427-L440

avishaiElmakies replied:
k

vasqu commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -696,32 +840,12 @@ def forward(
-        # required mask seq length can be calculated via length of past
-        mask_seq_length = past_key_values_length + seq_length
 
+        causal_attention_mask, attention_mask = self._update_causal_mask(
+            inputs_embeds, input_shape, past_key_values_length, attention_mask, output_attentions
+        )
```
Are we missing the handling of the cross-attention mask expansion (the encoder attention mask)? 

See for reference bart again:
https://github.com/huggingface/transformers/blob/78b2929c0554b79e0489b451ce4ece14d265ead2/src/transformers/models/bart/modeling_bart.py#L1286-L1321

I think we could just copy past it from bart directly, wdyt.

avishaiElmakies replied:
k

vasqu replied:
Thx for clarifying, wasn't aware of this. Still a bit weird that the attention implementation mostly follows bart then, but not that it matters.

avishaiElmakies replied:
Yeah, it was also weird to me, there was a refactor I tried to do, but I was told to revert it so it would still stay similar 😢

vasqu commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -615,6 +717,48 @@ def get_input_embeddings(self):
+                f"The provided attention mask has length {attention_mask.shape[1]}, but its length should be "
+                f"{mask_seq_length} (sum of the lengths of current and past inputs)"
+            )
+        if self._use_sdpa and not output_attentions:
```
Would need a check on `head_mask` to be None, else we would need to fal lback to the eager mask.

avishaiElmakies replied:
k

vasqu reviewed (commented):
Overall LGTM, needs a small check on the head mask too (eager fall back) and possibly the encoder attention mask hasn't been handled yet.

Commit:
if layer head mask default to eager

Commit:
if head mask is not none fall to eager

Commit:
[run-slow] opt

vasqu commented:
Thx LGTM, the failure on CI is not related to this PR.

vasqu commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -696,32 +841,12 @@ def forward(
-        # required mask seq length can be calculated via length of past
-        mask_seq_length = past_key_values_length + seq_length
 
+        causal_attention_mask, attention_mask = self._update_causal_mask(
```
Maybe a last nit: Is it necessary to keep the original attention mask over here?

avishaiElmakies replied:
the embedding layer takes the attention mask as input, there is another PR I have waiting for some feedback to change some of the logic there

https://github.com/huggingface/transformers/pull/33121#pullrequestreview-2283357050

vasqu replied:
Alright, makes sense. Just showed my unfamiliarity with opt as a model ^^

avishaiElmakies commented:
@amyeroberts would love a review, a slow-run test and hopefully a merge 🫡

molbap committed:
Clean up Unpack imports (#33631)

clean up Unpack imports

molbap committed:
Fix DPT /Dinov2 sdpa regression on main (#33660)

* fallback to eager if output attentions.

* fix copies

molbap committed:
handle dependency errors in check_imports (#33622)

* handle dependency errors in check_imports

* change log level to warning

chengchengpei committed:
add back self.max_position_embeddings = config.max_position_embeddings (#33550)

* add back self.max_position_embeddings = config.max_position_embeddings

* fix-copies

Isotr0py committed:
Fix Llava conversion for LlavaQwen2ForCausalLM with Clip vision tower (#33613)

fix llavaqwen2 model conversion

yonigozlan committed:
Uniformize kwargs for Udop processor and update docs (#33628)

* Add optional kwargs and uniformize udop

* cleanup Unpack

* nit Udop

gante committed:
Generation: deprecate `PreTrainedModel` inheriting from `GenerationMixin`  (#33203)

jiqing-feng committed:
Enable BNB multi-backend support (#31098)

* enable cpu bnb path

* fix style

* fix code style

* fix 4 bit path

* Update src/transformers/utils/import_utils.py

Co-authored-by: Aarni Koskela <akx@iki.fi>

* add multi backend refactor tests

* fix style

* tweak 4bit quantizer + fix corresponding tests

* tweak 8bit quantizer + *try* fixing corresponding tests

* fix dequant bnb 8bit

* account for Intel CPU in variability of expected outputs

* enable cpu and xpu device map

* further tweaks to account for Intel CPU

* fix autocast to work with both cpu + cuda

* fix comments

* fix comments

* switch to testing_utils.torch_device

* allow for xpu in multi-gpu tests

* fix tests 4bit for CPU NF4

* fix bug with is_torch_xpu_available needing to be called as func

* avoid issue where test reports attr err due to other failure

* fix formatting

* fix typo from resolving of merge conflict

* polish based on last PR review

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

* fix CI

* Update src/transformers/integrations/integration_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Update src/transformers/integrations/integration_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* fix error log

* fix error msg

* add \n in error log

* make quality

* rm bnb cuda restriction in doc

* cpu model don't need dispatch

* fix doc

* fix style

* check cuda avaliable in testing

* fix tests

* Update docs/source/en/model_doc/chameleon.md

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

* Update docs/source/en/model_doc/llava_next.md

Co-authored-by: Aarni Koskela <akx@iki.fi>

* Update tests/quantization/bnb/test_4bit.py

Co-authored-by: Aarni Koskela <akx@iki.fi>

* Update tests/quantization/bnb/test_4bit.py

Co-authored-by: Aarni Koskela <akx@iki.fi>

* fix doc

* fix check multibackends

* fix import sort

* remove check torch in bnb

* docs: update bitsandbytes references with multi-backend info

* docs: fix small mistakes in bnb paragraph

* run formatting

* reveret bnb check

* move bnb multi-backend check to import_utils

* Update src/transformers/utils/import_utils.py

Co-authored-by: Aarni Koskela <akx@iki.fi>

* fix bnb check

* minor fix for bnb

* check lib first

* fix code style

* Revert "run formatting"

This reverts commit ac108c6d6b34f45a5745a736ba57282405cfaa61.

* fix format

* give warning when bnb version is low and no cuda found]

* fix device assignment check to be multi-device capable

* address akx feedback on get_avlbl_dev fn

* revert partially, as we don't want the function that public, as docs would be too much (enforced)

---------

Co-authored-by: Aarni Koskela <akx@iki.fi>
Co-authored-by: Titus von Koeller <9048635+Titus-von-Koeller@users.noreply.github.com>
Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

tibor-reiss committed:
Fix error string after refactoring into get_chat_template (#33652)

* Fix error string after refactoring into get_chat_template

* Take suggestion from CR

Co-authored-by: Matt <Rocketknight1@users.noreply.github.com>

---------

Co-authored-by: Matt <Rocketknight1@users.noreply.github.com>

yonigozlan committed:
uniformize git processor (#33668)

* uniformize git processor

* update doctring

ArthurZucker committed:
Modular `transformers`: modularity and inheritance for new model additions (#33248)

* update exampel

* update

* push the converted diff files for testing and ci

* correct one example

* fix class attributes and docstring

* nits

* oups

* fixed config!

* update

* nitd

* class attributes are not matched against the other, this is missing

* fixed overwriting self.xxx now onto the attributes I think

* partial fix, now order with docstring

* fix docstring order?

* more fixes

* update

* fix missing docstrings!

* examples don't all work yet

* fixup

* nit

* updated

* hick

* update

* delete

* update

* update

* update

* fix

* all default

* no local import

* fix more diff

* some fix related to "safe imports"

* push fixed

* add helper!

* style

* add a check

* all by default

* add the

* update

* FINALLY!

* nit

* fix config dependencies

* man that is it

* fix fix

* update diffs

* fix the last issue

* re-default to all

* alll the fixes

* nice

* fix properties vs setter

* fixup

* updates

* update dependencies

* make sure to install what needs to be installed

* fixup

* quick fix for now

* fix!

* fixup

* update

* update

* updates

* whitespaces

* nit

* fix

* simplify everything, and make it file agnostic (should work for image processors)

* style

* finish fixing all import issues

* fixup

* empty modeling should not be written!

* Add logic to find who depends on what

* update

* cleanup

* update

* update gemma to support positions

* some small nits

* this is the correct docstring for gemma2

* fix merging of docstrings

* update

* fixup

* update

* take doc into account

* styling

* update

* fix hidden activation

* more fixes

* final fixes!

* fixup

* fixup instruct  blip video

* update

* fix bugs

* align gemma2 with the rest as well

* updats

* revert

* update

* more reversiom

* grind

* more

* arf

* update

* order will matter

* finish del stuff

* update

* rename to modular

* fixup

* nits

* update makefile

* fixup

* update order of the checks!

* fix

* fix docstring that has a call inside

* fiix conversion check

* style

* add some initial documentation

* update

* update doc

* some fixup

* updates

* yups

* Mostly todo gimme a minut

* update

* fixup

* revert some stuff

* Review docs for the modular transformers (#33472)

Docs

* good update

* fixup

* mmm current updates lead to this code

* okay, this fixes it

* cool

* fixes

* update

* nit

* updates

* nits

* fix doc

* update

* revert bad changes

* update

* updates

* proper update

* update

* update?

* up

* update

* cool

* nits

* nits

* bon bon

* fix

* ?

* minimise changes

* update

* update

* update

* updates?

* fixed gemma2

* kind of a hack

* nits

* update

* remove `diffs` in favor of `modular`

* fix make fix copies

---------

Co-authored-by: Lysandre Debut <hi@lysand.re>

ArthurZucker committed:
Fix CIs post merging modular transformers (#33681)

update

mnauf committed:
Fixed docstring for cohere model regarding unavailability of prune_he… (#33253)

* Fixed docstring for cohere model regarding unavailability of prune_head() methods

The docstring mentions that cohere model supports prune_heads() methods. I have fixed the docstring by explicitly mentioning that it doesn't support that functionality.

* Update src/transformers/models/cohere/modeling_cohere.py

---------

Co-authored-by: Lysandre Debut <hi@lysand.re>

gante committed:
Generation tests: update imagegpt input name, remove unused functions (#33663)

sizhky committed:
Improve Error Messaging for Flash Attention 2 on CPU (#33655)

Update flash-attn error message on CPU

Rebased to latest branch

gante committed:
Gemma2: fix config initialization (`cache_implementation`) (#33684)

umarbutler committed:
Fix ByteLevel alphabet missing when Sequence pretokenizer is used (#33556)

* Fix ByteLevel alphabet missing when Sequence pretokenizer is used

* Fixed formatting with `ruff`.

yonigozlan committed:
Uniformize kwargs for image-text-to-text processors (#32544)

* uniformize FUYU processor kwargs

* Uniformize instructblip processor kwargs

* Fix processor kwargs and tests Fuyu, InstructBlip, Kosmos2

* Uniformize llava_next processor

* Fix save_load test for processor with chat_template only as extra init args

* Fix import Unpack

* Fix Fuyu Processor import

* Fix FuyuProcessor import

* Fix FuyuProcessor

* Add defaults for specific kwargs kosmos2

* Fix Udop to return BatchFeature instead of BatchEncoding and uniformize kwargs

* Add tests processor Udop

* remove Copied from in processing Udop as change of input orders caused by BatchEncoding -> BatchFeature

* Fix overwrite tests kwargs processors

* Add warnings and BC for changes in processor inputs order, change docs, add BC for text_pair as arg for Udop

* Fix processing test fuyu

* remove unnecessary pad_token check in instructblip ProcessorTest

* Fix BC tests and cleanup

* FIx imports fuyu

* Uniformize Pix2Struct

* Fix wrong name for FuyuProcessorKwargs

* Fix slow tests reversed inputs align fuyu llava-next, change udop warning

* Fix wrong logging import udop

* Add check images text input order

* Fix copies

* change text pair handling when positional arg

* rebase on main, fix imports in test_processing_common

* remove optional args and udop uniformization from this PR

* fix failing tests

* remove unnecessary test, fix processing utils and test processing common

* cleanup Unpack

* cleanup

* fix conflict grounding dino

jmamou committed:
🚨🚨 Setting default behavior of assisted decoding (#33657)

dvrogozh committed:
tests: fix pytorch tensor placement errors (#33485)

This commit fixes the following errors:
* Fix "expected all tensors to be on the same device" error
* Fix "can't convert device type tensor to numpy"

According to pytorch documentation torch.Tensor.numpy(force=False)
performs conversion only if tensor is on CPU (plus few other restrictions)
which is not the case. For our case we need force=True since we just
need a data and don't care about tensors coherency.

Fixes: #33517
See: https://pytorch.org/docs/2.4/generated/torch.Tensor.numpy.html

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

amyeroberts commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -359,9 +362,106 @@ def forward(
+        position_ids: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        if output_attentions or layer_head_mask is not None:
+            logger.warning_once("""OPTModel is using SDPA attention, which currently does not support output_attentions=True.
+                                   failing back to eager attention. remove warning using attn_implementation="eager".""")
+
```
nit - formatting here is different from standard library patterns and elsewhere where the same warning is used 

```suggestion
            logger.warning_once(
                "OPTModel is using SDPA attention, which currently does not support output_attentions=True."
                'failing back to eager attention. remove warning using attn_implementation="eager".'
            )

```

amyeroberts reviewed (commented):
Thanks for adding this! 

Two things before merge: 
* OPT is a popular model, so we should add some benchmarking numbers and an example of how to use on the model's doc page. Here's a good example: https://huggingface.co/docs/transformers/v4.44.2/en/model_doc/gpt_neox#using-scaled-dot-product-attention-sdpa
* Rebasing on main should resolve the failing tests

ArthurZucker committed:
bump tokenizers, fix added tokens fast (#32535)

* update based on tokenizers release

* update

* nits

* update

* revert re addition

* don't break that yet

* fmt

* revert unwanted

* update tokenizers version

* update dep table

* update

* update in conversion script as well

* some fix

* revert

* fully revert

* fix training

* remove set trace

* fixup

* update

* update

NielsRogge committed:
[Pixtral] Improve docs, rename model (#33491)

* Improve docs, rename model

* Fix style

* Update repo id

ArthurZucker committed:
fix code quality after merge

bfineran committed:
HFQuantizer implementation for compressed-tensors library (#31704)

* Add compressed-tensors HFQuantizer implementation

* flag serializable as False

* run

* revive lines deleted by ruff

* fixes to load+save from sparseml, edit config to quantization_config, and load back

* address satrat comment

* compressed_tensors to compressed-tensors and revert back is_serializable

* rename quant_method from sparseml to compressed-tensors

* tests

* edit tests

* clean up tests

* make style

* cleanup

* cleanup

* add test skip for when compressed tensors is not installed

* remove pydantic import + style

* delay torch import in test

* initial docs

* update main init for compressed tensors config

* make fix-copies

* docstring

* remove fill_docstring

* Apply suggestions from code review

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

* review comments

* review comments

* comments - suppress warnings on state dict load, tests, fixes

* bug-fix - remove unnecessary call to apply quant lifecycle

* run_compressed compatability

* revert changes not needed for compression

* no longer need unexpected keys fn

* unexpected keys not needed either

* Apply suggestions from code review

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

* add to_diff_dict

* update docs and expand testing

* Update _toctree.yml with compressed-tensors

* Update src/transformers/utils/quantization_config.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* update doc

* add note about saving a loaded model

---------

Co-authored-by: George Ohashi <george@neuralmagic.com>
Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>
Co-authored-by: Sara Adkins <sara@neuralmagic.com>
Co-authored-by: Sara Adkins <sara.adkins65@gmail.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>
Co-authored-by: Dipika Sikka <ds3822@columbia.edu>
Co-authored-by: Dipika <dipikasikka1@gmail.com>

avishaiElmakies committed:
Update src/transformers/models/opt/modeling_opt.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

avishaiElmakies commented:
@amyeroberts is there some code to run those benchmarks?

avishaiElmakies committed:
Merge branch 'main' into spda_opt

avishaiElmakies commented:
@amyeroberts  ran benchmarks

used code from here https://github.com/huggingface/transformers/pull/31031

local resources - (L40S-45GB, PyTorch 2.4.0, OS Debian GNU/Linux 11

### Training 

|    batch_size |    seq_len |  Time per batch (eager - s)   |    Time per batch (sdpa - s) |  Speedup (%)   |  Eager peak mem (MB)   |    sdpa peak mem (MB) |  Mem saving (%)   |
|--------------:|-----------:|:------------------------------|-----------------------------:|:---------------|:-----------------------|----------------------:|:------------------|
|             1 |        128 | 0.047                         |                        0.037 | 26.360         | 1474.611               |               1474.32 | 0.019             |
|             1 |        256 | 0.046                         |                        0.037 | 24.335         | 1498.541               |               1499.49 | -0.063            |
|             1 |        512 | 0.046                         |                        0.037 | 24.959         | 1973.544               |               1551.35 | 27.215            |
|             1 |       1024 | 0.062                         |                        0.038 | 65.135         | 4867.113               |               1698.35 | 186.578           |
|             1 |       2048 | 0.230                         |                        0.039 | 483.933        | 15662.224              |               2715.75 | 476.718           |
|             2 |        128 | 0.045                         |                        0.037 | 20.455         | 1498.164               |               1499.49 | -0.089            |
|             2 |        256 | 0.046                         |                        0.037 | 24.027         | 1569.367               |               1551.35 | 1.161             |
|             2 |        512 | 0.045                         |                        0.037 | 20.965         | 3257.074               |               1698.35 | 91.778            |
|             2 |       1024 | 0.122                         |                        0.038 | 225.958        | 9054.405               |               2715.75 | 233.403           |
|             2 |       2048 | 0.464                         |                        0.067 | 593.646        | 30572.058              |               4750.55 | 543.548           |
|             4 |        128 | 0.045                         |                        0.037 | 21.918         | 1549.448               |               1551.35 | -0.123            |
|             4 |        256 | 0.044                         |                        0.038 | 18.084         | 2451.768               |               1698.35 | 44.361            |
|             4 |        512 | 0.069                         |                        0.037 | 84.421         | 5833.180               |               2715.75 | 114.791           |
|             4 |       1024 | 0.262                         |                        0.062 | 319.475        | 17427.842              |               4750.55 | 266.860           |
|             4 |       2048 | OOM                           |                        0.062 | Eager OOM      | OOM                    |               4750.55 | Eager OOM         |
|             8 |        128 | 0.044                         |                        0.037 | 18.436         | 2049.115               |               1697.78 | 20.694            |
|             8 |        256 | 0.048                         |                        0.036 | 32.887         | 4222.567               |               2715.75 | 55.484            |
|             8 |        512 | 0.153                         |                        0.06  | 154.862        | 10985.391              |               4750.55 | 131.245           |
|             8 |       1024 | 0.526                         |                        0.122 | 330.697        | 34175.763              |               8821.18 | 287.428           |
|             8 |       2048 | OOM                           |                        0.122 | Eager OOM      | OOM                    |               8821.18 | Eager OOM         |

### Inference

|    seq_len |    Per token latency eager (ms) |    Per token latency SDPA (ms) |    Speedup (%) |    Mem eager (MB) |    Mem BT (MB) |    Mem saved (%) |
|-----------:|--------------------------------:|-------------------------------:|---------------:|------------------:|---------------:|-----------------:|
|        128 |                          11.634 |                          8.647 |         34.546 |           717.676 |        717.674 |            0     |
|        256 |                          11.593 |                          8.86  |         30.851 |           742.852 |        742.845 |            0.001 |
|        512 |                          11.515 |                          8.816 |         30.614 |           798.232 |        799.593 |           -0.17  |
|       1024 |                          11.556 |                          8.915 |         29.628 |           917.265 |        895.538 |            2.426 |
|        128 |                          12.724 |                         11.002 |         15.659 |           762.434 |        762.431 |            0     |
|        256 |                          12.704 |                         11.063 |         14.83  |           816.809 |        816.733 |            0.009 |
|        512 |                          12.757 |                         10.947 |         16.535 |           917.383 |        918.339 |           -0.104 |
|       1024 |                          13.018 |                         11.018 |         18.147 |          1162.65  |       1114.81  |            4.291 |
|        128 |                          12.739 |                         10.959 |         16.243 |           856.335 |        856.483 |           -0.017 |
|        256 |                          12.718 |                         10.837 |         17.355 |           957.298 |        957.674 |           -0.039 |
|        512 |                          12.813 |                         10.822 |         18.393 |          1158.44  |       1158.45  |           -0.001 |
|       1024 |                          13.416 |                         11.06  |         21.301 |          1653.42  |       1557.19  |            6.18  |
|        128 |                          12.763 |                         10.891 |         17.193 |          1036.13  |       1036.51  |           -0.036 |
|        256 |                          12.89  |                         11.104 |         16.085 |          1236.98  |       1236.87  |            0.01  |
|        512 |                          13.327 |                         10.939 |         21.836 |          1642.29  |       1641.78  |            0.031 |
|       1024 |                          15.181 |                         11.175 |         35.848 |          2634.98  |       2443.35  |            7.843 |

updated model card opt.md as well

Commit:
update model card for opt

Commit:
add batch size to inference table

vasqu commented:
@avishaiElmakies Can you make the `[slow-run] opt` commit; it's needed to start slow runs. Great benchmarks btw!

Ci errors are not related.

Commit:
[slow-run] opt

avishaiElmakies commented:
@vasqu commited.

Yeah, the benchmarks look great. I was not expecting this much of an improvement when training

vasqu commented:
slow runs cc @amyeroberts

avishaiElmakies commented:
@amyeroberts it seems it have skipped the tests model, do you know what happend?

amyeroberts commented:
@avishaiElmakies I think it's because the message in the commit message is `[slow-run]` rather than `[run-slow]`. Could you try again with `[run-slow] opt`?

Commit:
[run-slow] opt

avishaiElmakies commented:
@amyeroberts sorry about that, commited with `[run-slow] opt`

vasqu commented:
The failures seem unrelated to me tbh. XLA should be tf stuff, no? @amyeroberts

amyeroberts commented:
@vasqu Yes, it should be TF stuff, although the TF files are touched in this PR so they should be passing too. Could you confirm if the tests are passing on main or not? 

cc @Rocketknight1 @ArthurZucker

avishaiElmakies commented:
@amyeroberts at the time those tests were failing in main as well

Rocketknight1 commented:
Confirmed that `test_xla_generate_contrastive` and `test_xla_generate_slow` are failing on `main` for me in TF - I think it's fine to just add skips to those tests in this PR for now, since it's not related to this PR.

vasqu commented:
Passing locally on my machine with a GPU ([tf_opt_logs.txt](https://github.com/user-attachments/files/17232562/tf_opt_logs.txt)) and not with a CPU ([tf_opt_logs_cpu.txt](https://github.com/user-attachments/files/17232603/tf_opt_logs_cpu.txt)).

Seems like a CPU issue then? FYI @Rocketknight1 
cc @amyeroberts

ArthurZucker commented:
Yep, we can add the wrapper `require_accelerator`!

vasqu commented:
@ArthurZucker There's a proper fix in #33903 since the GPU is just silently ignoring errors instead of throwing errors in TF even though we go beyond the possible positional embedding range.

avishaiElmakies commented:
@ArthurZucker @amyeroberts do i need to do anything or can this be merged?

ArthurZucker commented:
Sorry for the delay and thanks for the contribution! 🤗
